### ISSUE TITLE
loss NaN during training

### ISSUE THREAD

**Issue by saivinaypsv** [yuvrajkhanna/DTLN_AEC#1]
Epoch 00001: val_loss did not improve from inf
10000/10000 [==============================] - 4281s 428ms/step - loss: nan - val_loss: nan - lr: 0.0010
Epoch 2/100
10000/10000 [==============================] - ETA: 0s - loss: nan
Epoch 00002: val_loss did not improve from inf

**Comment by yuvrajkhanna** [comment]
There are lots of things I have seen make a model diverge and i am listing few of the reasons here, just make sure to cut them accordingly:
1. you may have an issue with the input data. Try calling assert not np.any(np.isnan(x)) on the input data to make sure you are not introducing the nan
2. Other numerical stability issues can exist such as there might be division by zero somewhere (altho i have added epsilon (1e-7) in the loss function to deal with that). but still make sure to check this.
3. Too high of a learning rate. You can often tell if this is the case if the loss begins to increase and then diverges to infinity. so in that case set the learning rate to some lower value. If none of these are working let me know

**Comment by saivinaypsv** [comment]
1 )            assert not np.any(np.isnan(nearend_signal_dat))
                assert not np.any(np.isnan(farend_signal_dat))
                assert not np.any(np.isnan(nearend_speech_dat))

                yield {"input_1": farend_signal_dat.astype('float32'), "input_2": 
                                                                  nearend_signal_dat.astype('float32')},nearend_speech_dat.astype('float32')

2 )  Added epsilon in loss function & all division operation

3) Learning rate 1e-3


*** Debugged model predictioin outputs in training time ##############
Batchsize = 2

########
[[0 0 0 ... -0.000946044922 -0.0022277832 -0.00390625]
 [-0.00335693359 -0.00381469727 -0.00448608398 ... 0.00143432617 0.00164794922 0.00109863281]]
 
  [[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]
##############

observed model output getting nan.

**Comment by yuvrajkhanna** [comment]
one output seems fine but the other resulted in nan. since u have already applied the assertion on input data, I am assuming your input is fine. so can you do one thing and check output of each layer to check which layer exactly is causing the problem, from there we can check on the issue.

**Comment by saivinaypsv** [comment]
First is ground truth data , this is proper... another is model prediction contains NaN.

I am trying to get each layer output in tensorflow.. but unfortunatey I could n't get much documentation about each layer output extraction ..I am not much familier in tensoflow. Can you give more insights about to print every layer output.

**Comment by yuvrajkhanna** [comment]
since this model isnt sequential, there isnt a really good way to print the values, what u can do is save the values in a text file after each layer where my model is written, instead of printing it. but first can you do these things:
1. reduce learning rate to 1e-5 and check if the problem is still there.
2. create the model, but instead of training it directly predict on the input to check if NaN still comes 
let me know the results of above 2

**Comment by saivinaypsv** [comment]
1). Treid with LR rate as 1e-5 , still model output is getting NaN


I have taken your github repository only .. but why its not working properly.. any bugs in repository ?

I am using synthetic data only (ICASS) , still getting nan...

**Comment by yuvrajkhanna** [comment]
I have written this code in 2020 for a college project at that time I trained this model on dataset which was available for DCASE task at that time and it worked perfectly fine for me so I am not sure.

**Comment by saivinaypsv** [comment]
Thanks.

I rewritten DTLN network in PyTorch , Now I'm abe train with Synthenic data   

1.   [farend_speech ,nearend_mic_signal ] as INPUTS
2.   [nearend_speech ] as TARGET

***********
Now I want traing with more database not with ICASSP synthetic data

Can you provied below datasets if poosible.. or else let me any opensource database avalaibale


farend_speech_path='E:/intern/codes/dataset/farend_speech/'
nearend_speech_path='E:/intern/codes/dataset/nearend_speech/'
RIR_path='E:/intern/codes/dataset/RIR/'

**Comment by saivinaypsv** [comment]
Found the NaN issue!!!

***In below loss function...... log10(zero) is -inf    if (num) numerator is zero loss getting -inf  *** 

 Need to add some epsilon in numerator also to get rid from nan.
-----------------------------------------------------------------------
    snr = tf.reduce_mean(tf.math.square(s_true), axis=-1, keepdims=True) / \
            (tf.reduce_mean(tf.math.square(s_true-s_estimate), axis=-1, keepdims=True)+1e-7)
            
        # using some more lines, because TF has no log10
        num = tf.math.log(snr) 
        
        denom = tf.math.log(tf.constant(10, dtype=num.dtype))
        loss = -10*(num / (denom))
        # returning the loss
        return loss

**Comment by saivinaypsv** [comment]
Another way to get rid from nan is add espilon for SNR value..

num = tf.math.log(snr+E)
